File: solutions/typescript/dnd-character/2/dnd-character.ts
function rollDice(): number {
    return Math.round(Math.random() * 6 + 1)
}

export class DnDCharacter {
  readonly strength = DnDCharacter.generateAbilityScore();
  readonly dexterity = DnDCharacter.generateAbilityScore();
  readonly constitution = DnDCharacter.generateAbilityScore();
  readonly intelligence = DnDCharacter.generateAbilityScore();
  readonly wisdom = DnDCharacter.generateAbilityScore();
  readonly charisma = DnDCharacter.generateAbilityScore();
  readonly hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
  
  public static generateAbilityScore(): number {
    return [...Array(4)].map(rollDice)
      .sort()
      .slice(1,4)
      .reduce((a, b) => a + b, 0);
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }
}

File: solutions/kotlin/dnd-character/1/src/main/kotlin/DndCharacter.kt
import kotlin.math.floor
import kotlin.random.Random

private fun rollDices(): Int {
    return Array(4) { Random.nextInt(1, 6) }
        .sortedDescending()
        .take(3)
        .sum()
}

class DndCharacter {
    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = 10 + modifier(constitution)

    companion object {
        fun ability() = rollDices()
        fun modifier(score: Int) = floor((score - 10) / 2.0).toInt()
    }
}
